### PR TITLE
Now `*CustomType` and `**CustomType` arguments are properly checked

### LINKED ISSUE #11144: keyword unpack isn't checked for anything that's not `dict`
# 😳
```py
from typing import Generic, TypeVar, Mapping
T = TypeVar("T")
T2 = TypeVar("T2")

class A(dict[T, T2]):
    ...

def foo(**i: float) -> None:
    ...

a: A[str, str]
d = {"a": "b"}

foo(**a)  # no error
foo(**d)  # error: Argument 1 to "foo" has incompatible type "**Dict[str, str]"; expected "float"
```
https://mypy-play.net/?mypy=latest&python=3.10&gist=1cbdc48546f3a1bc61476936c3dc9af7

Might be related to #11138

### LINKED ISSUE #11375: typecheck failure with parameterized iterator
`mypy-bug.py`
```python

from typing import *
import asyncio

T = TypeVar('T')

class Foo(Generic[T]):
    def __iter__(self) -> Iterator[Awaitable[T]]:
        return iter([])

async def bar(foo:Foo[T]):
    # No error
    await asyncio.gather(*iter(foo))  

    # mypy says:
    # No overload variant of "gather" matches argument type "Foo[T]"
    await asyncio.gather(*foo)

async def baz(foo:Foo):
    # also no error
    await asyncio.gather(*foo)
```

**To Reproduce**

```
➜  mypy mypy-bug.py
mypy-bug.py:17: error: No overload variant of "gather" matches argument type "Foo[T]"
mypy-bug.py:17: note: Possible overload variants:
mypy-bug.py:17: note:     def gather(coro_or_future1: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], coro_or_future2: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], coro_or_future3: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], coro_or_future4: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], coro_or_future5: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], coro_or_future6: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], *coros_or_futures: Union[Future[Any], Generator[Any, None, Any], Awaitable[Any]], loop: Optional[AbstractEventLoop] = ..., return_exceptions: bool = ...) -> Future[List[Any]]
mypy-bug.py:17: note:     def [_T1] gather(coro_or_future1: Union[Future[_T1], Generator[Any, None, _T1], Awaitable[_T1]], *, loop: Optional[AbstractEventLoop] = ..., return_exceptions: Literal[False] = ...) -> Future[Tuple[_T1]]
mypy-bug.py:17: note:     <9 more similar overloads not shown, out of 11 total overloads>
Found 1 error in 1 file (checked 1 source file)
```

**Expected Behavior**

The second `gather` call should typecheck just like the first one, as using `*foo` is essentially the same as `*iter(foo)`.

Note also that the type variable is required to reproduce the bug.

**Actual Behavior**

Mypy prints an error

**Your Environment**

- Mypy version used: mypy 0.812, also reproduces with mypy-0.910
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: Python 3.8.9
- Operating system and version: macOS 12.1

### PR DESCRIPTION
**Pull request by sobolevn**
The thing about this patch is that `ArgTypeExpander` needed `mapping_type` and `iterable_type` for checking `*` and `**` arguments.

But, this context was not provided. 
I've also fixed a bunch of related `TODO` items.

Closes #11144 
Closes  https://github.com/python/mypy/issues/11138
Closes #11375

### CODE REVIEW

File: mypy/argmap.py [hunk 1]
```diff
@@ -193,10 +200,13 @@ def expand_actual_type(self,
...
-                # Dict **arg.
-                # TODO: Handle arbitrary Mapping
+            elif (
+                isinstance(actual_type, Instance) and
+                len(actual_type.args) > 1 and
+                is_subtype(actual_type, self.context.mapping_type)
+            ):
+                # Only `Mapping` type can be unpacked with `**`.
+                # Other types will produce an error somewhere else.
                 return actual_type.args[1]
```
Comment (JelleZijlstra) [inline mypy/argmap.py:210]: Wouldn't this introduce a new bug if the generic args aren't the same as for Mapping? I guess that's #11138.

Comment (sobolevn) [reply mypy/argmap.py:210]: Yes, I am going to solve this as a part of #11138 🙂

File: test-data/unit/check-kwargs.test [hunk 2]
```diff
@@ -500,3 +500,57 @@ def g(x=1): pass
...
+    ...
+
+good1: Good
+good2: A[str, float]
+good3: B[str, float]
+foo(**good1)
+foo(**good2)
+foo(**good3)
+[out]
+main:29: error: Argument 1 to "foo" has incompatible type "**A[str, str]"; expected "float"
```
Comment (JukkaL) [inline test-data/unit/check-kwargs.test:551]: Can you update tests to use `# E: <message:` comments for error message?

File: mypy/argmap.py [hunk 3]
```diff
@@ -193,10 +203,16 @@ def expand_actual_type(self,
...
+                isinstance(actual_type, Instance) and
+                len(actual_type.args) > 1 and
+                is_subtype(actual_type, self.context.mapping_type)
+            ):
+                # Only `Mapping` type can be unpacked with `**`.
+                # Other types will produce an error somewhere else.
+                # TODO: this can be invalid, when generic type has:
+                # `class Some(Generic[A, B, C], Mapping[B, C]):`
+                # https://github.com/python/mypy/issues/11138
                 return actual_type.args[1]
```
Comment (JukkaL) [inline mypy/argmap.py:216]: Can you use `map_instance_to_supertype` to handle generic inheritance properly?

File: mypy/argmap.py [hunk 4]
```diff
@@ -162,16 +167,21 @@ def expand_actual_type(self,
...
-                if actual_type.type.fullname == 'builtins.list':
-                    # List *arg.
-                    return actual_type.args[0]
-                elif actual_type.args:
-                    # TODO: Try to map type arguments to Iterable
+            if isinstance(actual_type, Instance) and actual_type.args:
+                if is_subtype(actual_type, self.context.iterable_type):
+                    # TODO: this can be invalid, when generic type has:
+                    # `class Some(Generic[X, Y], Iterable[Y]):`
+                    # https://github.com/python/mypy/issues/11138
```
Comment (JukkaL) [inline mypy/argmap.py:178]: Can you use `map_instance_to_supertype` to work around the issue?

**Review by JukkaL — COMMENTED** [review]
Nice, this was a pretty bad gap in type checking. Left a few comments (not a full review).

File: mypy/argmap.py (hunk 4 above)
Comment (sobolevn) [reply mypy/argmap.py:178]: I've planned to properly work on it after this initial PR.
I've found several corner-cases that are out of scope of this piece of work.

But, if that's a problem I can just add it all at once 🙂

File: test-data/unit/check-kwargs.test (hunk 2 above)
Comment (sobolevn) [reply test-data/unit/check-kwargs.test:551]: Sure! Will do 👍

**Comment by sobolevn** [comment]
Ok, I am adding  https://github.com/python/mypy/issues/11138 support, it will require a couple of hours.

File: mypy/argmap.py (hunk 4 above)
Comment (sobolevn) [reply mypy/argmap.py]: Done! 👍

**Comment by sobolevn** [comment]
@hauntsaninja any ideas why `mypy_primer` fails?

**Comment by hauntsaninja** [comment]
Just pushed a fix... @msullivan 's hack for edgedb in mypy_primer ended up breaking

**Comment by sobolevn** [comment]
@hauntsaninja thank you! 🚀

**Review by hauntsaninja — APPROVED** [review]
Thanks, this is great!

Wish it was nicer to get the mapping and iterable types

**Comment by JukkaL** [comment]
This PR fixed an issue in an internal repo that I just encountered when testing the 0.920 release candidate!

**Comment by sobolevn** [comment]
Awesome! 🎉
